### ISSUE TITLE
Dask workers spilling to disk too early

### ISSUE THREAD

**Issue by KrishanBhasin** [dask/distributed#4061]
**What happened**:

I have a LocalCluster with ~50 workers. I use pure Dask Dataframe operations to create a dataframe with a large task graph (~3 million tasks).

When this dataframe is submitted to the LocalCluster (e.g. `ddf.to_csv()`), the compute finishes successfully, however at no point does RAM usage go above ~25% (the highest I have seen an individual worker go is ~35%). When I monitor the file system, I can see that `dask-worker-space` gets very large (>100GB).

I believe this is costing me lots of performance due to de/serialisation costs (I see tasks in the dashboard like e.g. `disk-write-read-csv`...).

I have set the dask config to as below (terminate at >100% as my data is skewed across workers - a single one breaking 100% isn't an issue):

```yaml
distributed:
  worker:
    memory:
      target: 0.7
      spill: 0.9
      pause: 1.0
      terminate: 1.2
```

When I have the cluster computing, I have attached a new client to it and run checks that the workers have been configured correctly (it appears they have):
```py

#returns 0.7 as expected
client.run(lambda dask_worker:dask_worker.memory_target_fraction)

#returns 0.9 as expected
client.run(lambda dask_worker:dask_worker.memory_spill_fraction)

#returns 1.0 as expected
client.run(lambda dask_worker:dask_worker.memory_pause_fraction)
```

I have DEBUG logs turned on, and also do not see any lines indicating spill-to-disk has started as per https://github.com/dask/distributed/blob/6a32208717c04aa9ccb4defdf253b8a00cab32b7/distributed/worker.py#L2667-L2669

If I set `worker.memory.target` to `False`, I continue to get low RAM usage (~25% per worker) and large `dask-worker-space`.

If I set `worker.memory.spill` to `False`, I continue to get low RAM usage (~25% per worker, apart from 1 of my workers which is at ~50%) and large `dask-worker-space`.

If I set `worker.memory.target` and `worker.memory.spill` to False, I see `dask-worker-space` remains empty and RAM usage steadily climbs. This unfortunately is not a solution for me, as my data doesn't wholly fit into memory, so this will eventually fail.

**What you expected to happen**:

- My workers to wait until they are at ~70% RAM util before `dask-worker-space` starts filling up.

**Minimal Complete Verifiable Example**:

I think colleague of mine got one using `dask.datasets.timeseries` - I will ask him to share it here.

```python
# Put your MCVE code here
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2.23.0 (seen in 2.19.0 also, possibly earlier)
- Python version: 3.7.7
- Operating System: Linux
- Install method (conda, pip, source): conda

**Comment by KrishanBhasin** [comment]
I've started by playing with the logging (adding a few lines of my own and upping the level to make sure I don't miss existing ones) and I can confirm that it is not entering into the below if statement branch.
```py
if self.memory_spill_fraction and frac > self.memory_spill_fraction: 
```
The `memory`, `frac`, `memory_spill_fraction` are all reporting "correctly" (that is, that they all indicate spilling should not be occurring).

I will keep digging, first by trying to create a more reliable MCVE

**Comment by mrocklin** [comment]
Thank you for your effort here Krishan

**Comment by KrishanBhasin** [comment]
@kylejn27's repro worked first time locally on my Mac so I'll poke around the internals over the course of the week.

**Comment by KrishanBhasin** [comment]
Latest update:

Creating the cluster with `processes=False` passed in leads to me being unable to replicate the issue (I get appropriate log messages: `Worker is at 71% memory usage. Start spilling data to disk` indicating the code is entering that branch)

**Comment by KrishanBhasin** [comment]
Just a further update on this:

On my large-task-graph usecase (mentioned in the original issue), when I don't use custom configuration, I still observe the RAM use hovering at 25-30%, but I occasionally get warnings from the Nanny about workers exceeding the 95% memory budget, and eventual failure via `KilledWorker`
e.g.
`KilledWorker: ("('getitem-73ddbaf4215b5913935e22a13e0aeae9', 92)", <Worker 'tcp://127.0.0.1:36281', name: 7, memory: 0, processing: 5057>)
`

**Comment by KrishanBhasin** [comment]
I'm wondering if this issue may be linked to my failure to get the colours on the dask dashboard's memory pane changing as I intended in https://github.com/dask/distributed/pull/3837?

**Comment by KrishanBhasin** [comment]
@mrocklin I have been playing around some more using my large task graph usecase from the original issue. I inserted some log lines that emit the values for the `zict` Buffer and have found out that:
- for my use case, the Buffer target is ~14GB (about correct for 0.7 * 20GB per worker)
- Zict's inspection of the buffer shows it is operating at ~97-99%, so it must be spilling before the worker's own `spill-to-disk` logic is used.
  - This is occuring while the worker still reports its own usage at ~25%

It appears to me that the issue is a disagreement between Zict and the worker's own self assessment of memory use. **Does this seem reasonable?**
I'm aware that my specific use case has (very) many unique keys, so I wonder if this is an avenue for me to explore further 

If so I will try to dig more, but I'm conscious of what I do not know (e.g. I do not know much about Zict beyond my skimming of the docs, and I do remember you briefly mentioning the different ways the workers can inspect memory in https://github.com/dask/distributed/issues/3799#issuecomment-629475854).

I would appreciate (without any obligation whatsoever) a quick ping if this seems reasonable or whether I'm running down a dead end.


I've got more details about how I obtained this information below:

<details> <summary>the diffs where I added log lines, and exactly what was emitted by them</summary>

The modified code looks like below from worker.py:

```python
# <the below is part of the Worker's __init__() function>
        elif self.memory_limit and (
            self.memory_target_fraction or self.memory_spill_fraction
        ):
            try:
                from zict import Buffer, File, Func
            except ImportError:
                raise ImportError(
                    "Please `python -m pip install zict` for spill-to-disk workers"
                )
            path = os.path.join(self.local_directory, "storage")
            storage = Func(
                partial(serialize_bytelist, on_error="raise"),
                deserialize_bytes,
                File(path),
            )
            target = int(float(self.memory_limit) * self.memory_target_fraction)
            self.data = Buffer({}, storage, target, weight)
            self.data.memory = self.data.fast
            self.data.disk = self.data.slow
        else:
            self.data = dict()
                 File(path),
             )
             target = int(float(self.memory_limit) * self.memory_target_fraction)
+            
+            logger.critical(f"the target set for zict is {target}")
+            
             self.data = Buffer({}, storage, target, weight)
             self.data.memory = self.data.fast
             self.data.disk = self.data.slow
```
which returns me `the target set for zict is 14440763473`.

```python 
    async def memory_monitor(self):
        """ Track this process's memory usage and act accordingly
        If we rise above 70% memory use, start dumping data to disk.
        If we rise above 80% memory use, stop execution of new tasks
        """
        if self._memory_monitoring:
            return
        self._memory_monitoring = True
        total = 0

        proc = self.monitor.proc
        memory = proc.memory_info().rss
        frac = memory / self.memory_limit


+        logger.critical(f"the zict looks like {self.data}")
+        logger.critical(f"the stuff in memory is {self.data.fast}")
+        logger.critical(f"the stuff on disk is {self.data.slow}")

        def check_pause(memory):
            frac = memory / self.memory_limit
            # Pause worker threads if above 80% memory use
            if self.memory_pause_fraction and frac > self.memory_pause_fraction:
                # Try to free some memory while in paused state
```
which returns me lines like
`2020-10-08 12:04:05,496 - distributed.worker - CRITICAL - the stuff in memory is <LRU: 14035039590/14440763473 on dict>`
`2020-10-08 12:04:05,513 - distributed.worker - CRITICAL - the stuff on disk is <Func: serialize_bytelist<->deserialize_bytes <File: /tmp/dask-worker-space/dask-worker-space/worker-l2ti_5t7/storage, mode="a", 574 elements>>`
`2020-10-08 12:04:05,616 - distributed.worker - CRITICAL - the zict looks like Buffer<<LRU: 14035039590/14440763473 on dict>, <Func: serialize_bytelist<->deserialize_bytes <File: /tmp/dask-worker-space/dask-worker-space/worker-l2ti_5t7/storage, mode="a", 574 elements>>> `

A small aside - I direct distributed to use `/tmp/` as scratch space, and it looks like it creates a `dask-worker-space` directory inside `dask-worker-space`, inside `/tmp`.

</details>

**Comment by KrishanBhasin** [comment]
In summary:
- It appears that the Workers' two methods of identifying memory use have been strongly disagreeing in my use case.
`dask.sizeof()` seems to massively overestimate memory use compared to `psutil`.
- I solved my issue by setting `memory.target` to `False` in the configuration file, relying on `psutil`'s memory monitoring to trigger a spill-to-disk at my `memory.spill` target.
  - This initially did not work due to how `False` was propagating through to Zict and how it decided whether to spill-to-disk (it was spilling everything immediately), which has since been resolved via #4156 

I am not clear on why `dask.sizeof()` is massively overestimating my memory use, and am not able to dedicate time to investigating this further at the moment.

I'm going to close this issue as it has been (mostly) understood, and I hope this summary helps any future users of Dask/Distributed who encounter similar behaviour!

(if closing the issue is the wrong action, please feel free to reopen)
